crates/kumod/src/smtp_server.rs

```diff
@@ -2264,6 +2280,177 @@ impl SmtpServerSession {
+                                when: Utc::now(),
+                            });
+                            self.read_buffer.extend_from_slice(&data[0..size]);
+                        }
+                    }
+                }
+                _ = self.shutdown.shutting_down() => {
+                    return Err(anyhow!("shutting down"));
+                }
+            };
+            let header = HeaderResult::parse(&self.read_buffer);
+            if header.is_complete() {
+                break header;
+            }
+        };
+
+        let mut addr: Option<IpAddr> = None;
+        let mut port: Option<u16> = None;
+        let mut dest_addr: Option<IpAddr> = None;
+        let mut dest_port: Option<u16> = None;
+        match header {
+            HeaderResult::V1(Ok(header)) => {
+                let addresses = header.addresses;
+                match addresses {
+                    ppp::v1::Addresses::Tcp4(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
+                        dest_addr.replace(addresses.destination_address.into());
+                        dest_port.replace(addresses.destination_port);
+                    }
+                    ppp::v1::Addresses::Tcp6(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
+                        dest_addr.replace(addresses.destination_address.into());
+                        dest_port.replace(addresses.destination_port);
+                    }
+                    _ => {}
+                }
+            }
+            HeaderResult::V2(Ok(header)) => {
+                let addresses = header.addresses;
+                match addresses {
+                    ppp::v2::Addresses::IPv4(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
+                        dest_addr.replace(addresses.destination_address.into());
+                        dest_port.replace(addresses.destination_port);
+                    }
+                    ppp::v2::Addresses::IPv6(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
+                        dest_addr.replace(addresses.destination_address.into());
+                        dest_port.replace(addresses.destination_port);
+                    }
+                    _ => {}
+                }
+            }
+            HeaderResult::V1(Err(error)) => {
+                return Err(anyhow!("proxy protocol v1 parsing error: {error}"))
+            }
+            HeaderResult::V2(Err(error)) => {
+                return Err(anyhow!("proxy protocol v2 parsing error: {error}"))
+            }
+        }
+        self.read_buffer.clear();
```

Comment:
This may be a slightly moot comment, due to how SMTP is phrased, but this unilateral buffer clear has the potential to discard unread data.

Let's say that a slightly non-conforming client connects and immediate sends `EHLO domain` after the proxy protocol header, without waiting for the server to send the banner.  If the proxy header is relatively short, `read_buffer` now contains a fragment of that `EHLO` and when we reach this point the buffer clear will discard it.

Subsequently, when we start processing lines for real, we'll consider that first command to be nonsense.

What we should do here is consume only the bytes that were part of the recognized proxy header.  The ppp types make that look a bit awkward, but I think that you can read the `.len()` from either:

* https://docs.rs/ppp/latest/ppp/v1/struct.Header.html#structfield.header
* https://docs.rs/ppp/latest/ppp/v2/struct.Header.html#structfield.header

Armed with that consumed length, we can then `drain` out that many bytes from the read buffer, something like: `self.read_buffer.drain(0..consumed_len)`.

```suggestion
        self.read_buffer.drain(0..consumed_bytes);
```